File: java/src/Buscador.java
import java.util.ArrayList;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicIntegerArray;

public class Buscador {
    public static boolean sequencial(GrafoBusca grafoBusca) {
        ArrayList<Integer> visitados = new ArrayList<>();
        ArrayList<Integer> visitar = new ArrayList<>();

        visitar.add(grafoBusca.getInicio());

        while(!visitar.isEmpty()) {
            ArrayList<Integer> proximosVisitar = new ArrayList<>();

            for (Integer verticeVisitando : visitar) {
                if(grafoBusca.getGrafo().getValorVertice(verticeVisitando).equals(grafoBusca.getValorBusca())) {
                    return true;
                }

                visitados.add(verticeVisitando);

                for (Integer verticeAdjacente : grafoBusca.getGrafo().getVerticesAdjacentes(verticeVisitando)) {
                    proximosVisitar.add(verticeAdjacente);
                }
            }

            visitar = proximosVisitar;
        }

        return false;
    }

    public static boolean concorrente(GrafoBusca grafoBusca, int nThreads) throws InterruptedException, ExecutionException {
        AtomicIntegerArray visitados = new AtomicIntegerArray(grafoBusca.getGrafo().size());
        ArrayList<Integer> visitar = new ArrayList<>();

        ExecutorService executorService = Executors.newFixedThreadPool(nThreads);
        AtomicBoolean encontrou = new AtomicBoolean();

        visitar.add(grafoBusca.getInicio());

        while(!visitar.isEmpty() && !encontrou.get()) {
            ArrayList<Future<ArrayList<Integer>>> futures = new ArrayList<>();

            for (Integer verticeVisitando : visitar) {
                futures.add(executorService.submit(() -> {
                    ArrayList<Integer> proximosVisitar = new ArrayList<>();
                    
                    if(visitados.get(verticeVisitando) == 1 || encontrou.get()) {
                        return proximosVisitar;
                    }

                    if(grafoBusca.getGrafo().getValorVertice(verticeVisitando).equals(grafoBusca.getValorBusca())) {
                        encontrou.set(true);
                        return proximosVisitar;
                    }

                    for (Integer verticeAdjacente : grafoBusca.getGrafo().getVerticesAdjacentes(verticeVisitando)) {
                        if(visitados.get(verticeAdjacente) == 1) {
                            continue;
                        }
                        proximosVisitar.add(verticeAdjacente);
                    }

                    return proximosVisitar;
                }));
            }

            visitar = new ArrayList<>();

            for (Future<ArrayList<Integer>> future : futures) {
                visitar.addAll(future.get());
            }
        }

        executorService.shutdownNow();

        return encontrou.get();
    }
}

File: cpp/Buscador.cpp
#include "Buscador.h"
#include <omp.h>
#include <algorithm>
#include <queue>

bool Buscador::sequencial(const GrafoBusca& grafoBusca) {
    std::vector<bool> visitados(grafoBusca.getGrafo().size(), false);
    std::vector<int> visitar = {grafoBusca.getInicio()};

    while (!visitar.empty()) {
        std::vector<int> proximosVisitar;

        for (int verticeVisitando : visitar) {
            if (grafoBusca.getGrafo().getValorVertice(verticeVisitando) == grafoBusca.getValorBusca()) {
                return true;
            }

            visitados[verticeVisitando] = true;

            for (int verticeAdjacente : grafoBusca.getGrafo().getVerticesAdjacentes(verticeVisitando)) {
                if (!visitados[verticeAdjacente]) {
                    proximosVisitar.push_back(verticeAdjacente);
                }
            }
        }

        visitar = proximosVisitar;
    }

    return false;
}

bool Buscador::paralelo(const GrafoBusca& grafoBusca, int nThreads) {
    std::vector<bool> visitados(grafoBusca.getGrafo().size(), false);
    std::vector<int> visitar = {grafoBusca.getInicio()};
    bool encontrado = false;

    while (!visitar.empty() && !encontrado) {
        std::vector<int> proximosVisitar;

        #pragma omp parallel num_threads(nThreads)
        {
            std::vector<int> locaisProximosVisitar;

            #pragma omp for nowait
            for (int i = 0; i < visitar.size(); ++i) {
                
                int verticeVisitando = visitar[i];

                if (grafoBusca.getGrafo().getValorVertice(verticeVisitando) == grafoBusca.getValorBusca()) {
                    #pragma omp critical 
                    {
                        encontrado = true;
                    }
                }

                #pragma omp critical 
                {
                    visitados[verticeVisitando] = true;
                }

                for (int verticeAdjacente : grafoBusca.getGrafo().getVerticesAdjacentes(verticeVisitando)) {
                    if (!visitados[verticeAdjacente]) {
                        locaisProximosVisitar.push_back(verticeAdjacente);
                    }
                }

            }

            #pragma omp critical 
            {
                proximosVisitar.insert(proximosVisitar.end(), locaisProximosVisitar.begin(), locaisProximosVisitar.end());
            }
        }

        #pragma omp flush(encontrado)
        if (encontrado) break;

        // Remove duplicatas dos próximos vértices a visitar
        std::sort(proximosVisitar.begin(), proximosVisitar.end());
        proximosVisitar.erase(std::unique(proximosVisitar.begin(), proximosVisitar.end()), proximosVisitar.end());

        visitar = proximosVisitar;
    }

    return encontrado;
}
